assembly/src/ast/mod.rs

```diff
@@ -290,22 +296,30 @@ impl ModuleAst {
     /// A module consists of internal and exported procedures but does not contain a body.
     pub fn new(
         local_procs: Vec<ProcedureAst>,
+        reexported_procs: Vec<ProcReExport>,
```

Comment:
We should consider merging these two fields. The locally defined procedures and the re-exported ones together define the interface for the module, and the AST should reflect this. This would also make the parser solely responsible for dealing with name clashes (be it by shadowing or throwing an error), rather than the assembler itself.

I also think they should be made a set rather than a vector, so that name clashes aren't allowed at all.

Reply:
I mean something like this:

```
struct ProgramAst {
    // local_procs: Vec<ProcedureAst>,
    // reexported_procs: Vec<ProcReExport>,
    procs: BTreeSet<ProcType>  // Replaces the previous two lines
    ...
}

enum ProcType {
    LocalProc(ProcedureAst),
    ReExportProc(ProcReExport)
}
```

If reexported procedures are stored separately, then anyone who uses `ProgramAst` will need to figure which scoping rule to implement (Is shadowing is allowed? If so, which procedure shadows which? Etc.).

Instead, if the parser collapses all the visible procedures into the same collection we eliminate that confusion: There is only one collection, so only one place to look for the procedure, and it's a set, so the procedure we find is unique. In effect, we implement the scoping rules once and for all in the parser, and never need to worry about them anywhere else.

Reply:
Actually, this isn't quite right. It would still be possible to have two procedures with the same name in this collection if one is locally defined and the other is a reexport.

Something like this, maybe:

```
struct ProgramAst {
    // local_procs: Vec<ProcedureAst>,
    // reexported_procs: Vec<ProcReExport>,
    procs: BTreeMap<ProcedureName, ProcType>  // Replaces the previous two lines
    ...
}
```

This isn't quite right either, because we use ProcedureId to find the correct procedure, but it's something like this I mean: A single collection that prevents name clashes by design, so that there is no confusion as to which procedure is in scope.

Reply:
This isn't actually as good a suggestion as I initially thought.

The suggestion works well when another program imports the module we're dealing with, but when the module calls it's own procedures we actually do want them separated, because we use different instructions for calls to local procedures and calls to imported ones.

Probably best to keep them separate in the AST, and combine them when they are exported, just like you're currently doing.